keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,539 @@
+
+## Proposal
+
+Volume monitoring is the main focus of this proposal. Reactions are not in the scope of this proposal.
+
+- If a volume provisioned by the CSI driver is deleted, we need to mark the corresponding PVC to inform users.
+
+- If mounting error occurs, we need to mark the PVC.
+
+- If other errors occur, we also need to mark the PVC.
+
+- If volume errors are detected by other components deployed by users, we need to mark the PVC.
+
+The main architecture is shown below:
+
+![pv health monitor architecture](./pv-health-monitor.png)
+
+The following areas will be the focus of this proposal at first:
+
+- The health condition checking of volumes themselves. For example, whether the volume is deleted, whether the usage is reaching the threshold, and so on.
+- Mounting conditions checking.
+- Other errors that could affect the usability of the volume.
+
+In addition, a service will be created to receive volume health condition reports from other compoments implemented and deployed by users.
+
+Three main parts are involved here in the architecture.
+
+- API changes: The Status field in the PVC will be used to mark volumes if they are unhealthy.
+- External Controller:
+  - Trigger controller RPC to check the health condition of network volumes themselves for network storage.
+  - Watch for node failure events for both network and local storage.
+  - Create HTTP(RPC) service to receive PVs health condition reports.
+
+- External Agent:
+  - Trigger node RPC to check PVs’ mounting conditions for network storage.
+  - Trigger node RPC to check local volumes health condition for local storage;
+    - Note that currently we do not have CSI support for local storage. As a workaround, we may check the local volumes directly by the agent at first, and then move the checks to RPC interfaces when it is ready.
+
+
+## Implementation
+
+### API changes
+
+The Status field in the PVC will be used to mark volumes if they are unhealthy.
+
+```
+// PersistentVolumeClaimHealthConditionType defines the health condition of PV claim.
+// Valid values are "VolumeNotFound", "VolumeUnMounted", "VolumeOutOfCapacity", or "VolumeIOError".
+type PersistentVolumeClaimHealthConditionType string
+
+// These are valid health conditions of PVC
+const (
+        // PersistentVolumeClaimVolumeNotFound - Volume not found
+        PersistentVolumeClaimVolumeNotFound PersistentVolumeClaimHealthConditionType = "VolumeNotFound"
+        // PersistentVolumeClaimVolumeUnMounted - Volume not mounted
+        PersistentVolumeClaimVolumeUnMounted PersistentVolumeClaimHealthConditionType = "VolumeUnMounted"
```

Comment:
Does Kubernetes actually need explicit error conditions, or can it just pass through the CSI error codes?

Adding new error codes will be hard if it requires a K8s API change every time. And there's no way to surface any of the "custom" error codes that drivers may have.

Reply:
We need to at least show that the volume in unhealthy in the status field.  Maybe we can just have simplified health conditions like these:
- HealthFailure
- HeathWarning
- HealthUnknown
All the detailed error codes can be reported by the driver and will be included in the message field?